### ISSUE TITLE
getting an object of the List<T> throws casting exception after hot restart

### ISSUE THREAD

**Issue by michaelgobbers** [hivedb/hive#234]
**Steps to Reproduce**
Because Hive is trying to cast List<dynamic> to a given List<T> a cast exception is happening inside of hive. The workaround is to just fetch a List without supplying the concrete type and then do a cast() or a fromList(). I do think that this is an actual bug as the behaviour for this is different when you just added the data to the cache as opposed to the get after a hot reload. In both cases it should behave the same in my opinion. This is a source of bugs that might go unnoticed at first.

I know that there have been issues reported like this and the response was that this is because of dart limitations however I do feel that this is still a bug and that this is fixable. A possible solution would be (I haven't confirmed this) would be to use the dart mirrors library (https://api.dart.dev/stable/2.7.1/dart-mirrors/dart-mirrors-library.html) that provides some basic reflection to cover this case. 

**Code sample**
```
var box = openBox<List<int>>("some box name");
box.put("some key", [1, 2, 3]);
box.get("some key"); //This works without an exception

//perform hot restart
var box = openBox<List<int>>("some box name");
box.get("some key"); //This throws an exception
```

**Version**
 - Platform: iOS, Android
 - Flutter version: 1.12.13+hotfix.8
 - Hive version: 1.4.0+1

**Comment by leisim** [comment]
> A possible solution would be (I haven't confirmed this) to use the dart mirrors library

Flutter does not support the mirrors library. I'm also not happy with this situation and I think it is a serious flaw of the Dart language but unfortunately, there is no workaround currently. Because of that, I don't consider it a problem with Hive.

**Comment by andreidiaconu** [comment]
At the very least, hive should behave the same across scenarios. In the described scenario, it works just fine until you restart the app -- which was quite surprising.

**Comment by cntseesharp** [comment]
List<Country>() is List - true;

With this comparison it's possible to wrap "value as T" into 
```
if(value is List)
{
return List<T>.from(value);
}
```

**Comment by leisim** [comment]
> At the very least, hive should behave the same across scenarios.

The behavior is documented and it is not possible to check if you provide a `List<Something>` or `List`.

> With this comparison it's possible to wrap "value as T" into

But you need to get `T` and I don't think it is possible. Also what about `Map<X, Y>` or `YourClass<A, B, C, D>`?

**Comment by michaelgobbers** [comment]
Thanks for your fast reply! I'm just throwing an idea here but my assumption here is that the difference in behaviour after a hot reload is the fact that we don't have any type information stored of what the box actually contains right? We lose that after a hot restart. 

If that's the case, wouldn't it be a possibility to also persist this type information and use that information when loading the data again when reading from disk?

**Comment by leisim** [comment]
> wouldn't it be a possibility to also persist this type information

I don't think so. To my knowledge it is impossible to get `T` from a `List<T>`. Also I don't think it is possible to persist a complex type like `List<T>` or `Map<X, Y>`.

Maybe one option would be the following check when a box is opened:

```dart
Box<T> openBox<T>() {
  if ((T is List && T != List) || (T is Map && T != Map)) {
    throw 'Lists and maps are not allowed to have a generic argument';
  }
}
```

I'm not entirely sure whether this code works in VM and browser but I think it does.

What do you think?

**Comment by cntseesharp** [comment]
@leisim i agree, wrong assumption. Partially. Imo reflection can allow us to do 
```
Activator.createInstance(T).from(value);
```
Even in case this throws build error it can be wrapped in 
```
(Activator.createInstance(T) as dynamic).from(value);
```
Which is a dirty hack, but it's gonna work

**Comment by michaelgobbers** [comment]
Sorry for the late reply. I kindof lost track of this one. 
I think this should work yes and at least prevent the user from using the library in such a way.
This will not cover all of the cases such as user defined types with generics, however you would already impact a large amount of consumers of this library because I believe this is the typical mistake people would make at first.

**Comment by MoacirSchmidt** [comment]
Does that mean that I cannot have a Box<List<int>> box?

**Comment by GursheeshSingh** [comment]
I am facing this error too!!
Couldn't find any mentions in Doc and was hard to find this issue.

I am also trying to use Box<List<T>>

But it keeps throwing error

What is the solution?

**Comment by TheMisir** [comment]
> I am facing this error too!!
> Couldn't find any mentions in Doc and was hard to find this issue.
> 
> I am also trying to use Box<List>
> 
> But it keeps throwing error
> 
> What is the solution?

As I understand from previous conversation, the issue is related to dart SDK rather than hive. So, I would suggest using custom class and adapter to convert data to/from binary. Check [type adapter](https://docs.hivedb.dev/#/custom-objects/type_adapters) section from docs.

**Comment by andreidiaconu** [comment]
@GursheeshSingh 
Here are 2 Solutions, according to what you want to achieve. Let's assume you want to store `List<Person>` in a `Box<List<Person>>`.
### Solution 1
First consider if you can use the Box directly, without the List. Having `Box<Person>` is in many cases what you wanted to do anyway. You can get all values using `box.values()` which returns an `Iterable<Person>` that you can work with. You should view a Box as a collection -- so you are trying to create a collection of lists of objects
### Solution 2
Wrap your list in something that does not use generics and use that with a custom adapter. This would make your box be a `Box<People>` and then People is a class that contains property `List<Person> list;`. You then use an adaptor and the generator, which deal with reconstructing the object the right way for you.
